Custom Kinetic Accelerators

BurgerLUA opened the pull request:
# Overview
This is honestly one of my biggest and most ambitious projects. I hope people are happy with this.

Custom kinetic accelerators are special, customizable mining weapons that are meant to give a sense of progression, and a sense of pride and accomplishment while mining. Each custom KA is made up of 3 or 4 parts:

- The Frame (5 to choose from)
- The Cell (5 to choose from, +1 traitor)
- The Barrel (5 to choose from, +1 traitor)
- The Upgrade Chip (7 to choose from, +1 traitor)

The sprites change with each addon, they are truly dynamic and there are so many combinations of parts that a miner can have. You can have several different builds for each desired mining style, it's quite a robust system.

You can have a KA that shoots slowly, but delivers high-penetrating shots that go through several walls of rock. You can have a KA that shoots REALLY fast, but precisely destroys 1 rock at a time. You can have an absolute canon of a beast, that destroys rocks in a 3 tile radius around it.

The parts can be obtained via research, via abandoned crates, or via RNG in the cargo warehouse. There are some custom KAs that spawn on the raider's ship, on the mercs ship, and a laser one on the merc's headquarters. That laser one can also be bought from the traitor uplink as well.

ERT get a class 4 KA for use in special operations, if they choose to use it.

An overview of sprites here: https://imgur.com/a/x9VvVVa

BotBOREALIS commented:
Icons with diff:
<details><summary>icons/obj/kinetic_accelerators.dmi</summary>

Key | Old | New | Status
--- | --- | --- | ---
frame01|![]()|![frame01](http://images.aurorastation.org/images/c08ccdad60ad3e649f7624a88f54fe64.png)|Created
frame02|![]()|![frame02](http://images.aurorastation.org/images/755a55bf1f2c9320c9cd267e33ea668d.png)|Created
frame03|![]()|![frame03](http://images.aurorastation.org/images/b7eaab5ee93c561e74118c994f08afaf.png)|Created
frame04|![]()|![frame04](http://images.aurorastation.org/images/592a4a7190c6bfa089ff1f50098b03e4.png)|Created
frame05|![]()|![frame05](http://images.aurorastation.org/images/d8bd50d80321cd843dec2264b30c29d8.png)|Created
barrel01|![]()|![barrel01](http://images.aurorastation.org/images/8ef8819c9f04998d5bd1a1a15215cb17.png)|Created
barrel02|![]()|![barrel02](http://images.aurorastation.org/images/5a5474c4b38ffca732fd9b39a4ad8b82.png)|Created
barrel03|![]()|![barrel03](http://images.aurorastation.org/images/b6012e73a1e1c98aa91e49df530bfd49.png)|Created
barrel04|![]()|![barrel04](http://images.aurorastation.org/images/f952bb9dd904bbfd078198750847b367.png)|Created
barrel05|![]()|![barrel05](http://images.aurorastation.org/images/ce6d01e0d2e92d794c2ff1f312101cfb.png)|Created
cell01|![]()|![cell01](http://images.aurorastation.org/images/6088ff498319f4913156215dc048869e.png)|Created
cell02|![]()|![cell02](http://images.aurorastation.org/images/3b3a394d63000653fd840d2749e3a455.png)|Created
cell03|![]()|![cell03](http://images.aurorastation.org/images/275a86f54f685e670c69d6666a23c7f3.png)|Created
cell04|![]()|![cell04](http://images.aurorastation.org/images/2ef52830510a763c40413f6ee6682579.png)|Created
cell05|![]()|![cell05](http://images.aurorastation.org/images/c9a82896be424ebbf1bccc2f7fb0027d.png)|Created
upgrade_chip|![]()|![upgrade_chip](http://images.aurorastation.org/images/ff34d102482b6ce3c07c2a0f9fdcbdc1.png)|Created
|![]()|![](http://images.aurorastation.org/images/390497788cbe7b4a69d67ac2ed0b9a4f.png)|Created
barrel_laser|![]()|![barrel_laser](http://images.aurorastation.org/images/efa5c346c3ea0a8a5f6dcd9a54adc2c6.png)|Created
upgrade_chip_illegal|![]()|![upgrade_chip_illegal](http://images.aurorastation.org/images/1ce441c313632b7c26d7bd2e91c9bacf.png)|Created
cell_illegal|![]()|![cell_illegal](http://images.aurorastation.org/images/e049f191f7a156428391a32582b14cc1.png)|Created
kineticgun_lh|![]()|![kineticgun_lh](http://images.aurorastation.org/images/176d6174989d41278780c77a26797aa2.png)|Created
kineticgun_lh[1]|![]()|![kineticgun_lh[1]](http://images.aurorastation.org/images/86cc280ab9374687bc4811e5fec042e4.png)|Created
kineticgun_lh[2]|![]()|![kineticgun_lh[2]](http://images.aurorastation.org/images/e18c73b0687eb590c0c9a02ac3a3828b.png)|Created
kineticgun_lh[3]|![]()|![kineticgun_lh[3]](http://images.aurorastation.org/images/6889f7c3ac52a940f43ac0d55b5b8b05.png)|Created
kineticgun_rh|![]()|![kineticgun_rh](http://images.aurorastation.org/images/941f8692f1061ff08ece2bcacc70abf8.png)|Created
kineticgun_rh[1]|![]()|![kineticgun_rh[1]](http://images.aurorastation.org/images/79c85164540871ce93f972ce03554bc3.png)|Created
kineticgun_rh[2]|![]()|![kineticgun_rh[2]](http://images.aurorastation.org/images/6a37caad89bcf71e1c3be63fc2c8a406.png)|Created
kineticgun_rh[3]|![]()|![kineticgun_rh[3]](http://images.aurorastation.org/images/5892e4be6addc443329536182991d46b.png)|Created
</details>

Pacmandevil commented:
inb4 someone complains because the stock KPA winds up being like 1% less effective

OneOneThreeEight commented:
wouldn't that be weird, that an un-upgraded device isn't as effective as an upgraded one? nah, totally worth the outrage

OneOneThreeEight commented on code/modules/custom_ka/barrels.dm:
```diff
@@ -0,0 +1,96 @@
+
+/obj/item/custom_ka_upgrade/barrels/barrel02
+	name = "professional core KA power converter"
+	build_name = "core XXC 'Profesional'"
```
Misspell of 'professional.' Not to be nitpicky.

Lohikar commented on code/modules/custom_ka/frames.dm:
```diff
@@ -0,0 +1,127 @@
+	range = 5
+
+	var/pressure_decrease = 0.25
+	var/list/hit_overlays = list()
```
This list is never written to, only read.

Alberyk commented:
Some issues:
-the illegal ka is pretty much a laser weapon, it is pointless to give them to the mercs at round start when they already have access to laser carbines
-the names are way too long and they clash with the other gun names

Alberyk commented:
Also, what is exactly being changed on the maps?

BurgerLUA commented:
I added a single wrench to the main level in mining, as wrenches are required to modify custom KAs.
I also added 2 random, custom KAs on the raider shuttle, and 3 random, custom KAs on the merc shuttle, and 3 level 4 custom KAs in the ERT specialist area.

Because mapmerge is mapmerge, it did a lot of file-related changes.

Alberyk commented:
Pretty sure that mining already has a toolbox in the mining bay, and I don't see the need for mercs or raiders to have access to those, they already have weapons and mining drills. And why does even ert needs them?

BurgerLUA commented:
Reason: I want people to try them out.

Excuse: ERT only have 2 things of mining equipment, and it's the hardsuit mounted drill. I just think it would be pretty neat for ERT to get non-plebeian equipment if a dig-related situation arises. Raiders would likely have access to them given how they steal, and given how the KAs they steal are randomized. It would be possible for them to randomize a KA that doesn't work. I'll probably get rid of the mercs KA since they already have diamond drills.

BurgerLUA commented:
I can't perform a mapmerge on the centcomm level because map merge 1 is not smart enough to do it, for some reason. Waiting until there's a fix to this problem until I attempt merging.

BurgerLUA commented:
Fixed the merge conflict. Centcomm and map merge fixer do not like behaving, so centcomm is now left unchanged, meaning that ERT and Mercs no longer get custom KAs.

LordFowl commented on code/game/objects/items/weapons/manuals.dm:
```diff
@@ -1367,4 +1367,84 @@
\ No newline at end of file
+			"}
+
+/obj/item/weapon/book/manual/brainwashing
```
What does brainwashing have to do with pKAs?

BurgerLUA replied:
It was leftover from an poorly done devmerge on my part.

LordFowl commented on code/modules/custom_ka/barrels.dm:
```diff
@@ -0,0 +1,96 @@
+/obj/item/custom_ka_upgrade/barrels/barrel05
+	name = "experimental core KA power converter"
+	build_name = "'Experimental'"
+	desc = "A very experimental kinetic accelerator energy converter. Not much is known about this thing, other than it kicks like a muel and stings like an e-sword."
```
Mule. And I've never heard of e-swords being notable for their stinginess.

LordFowl commented on code/modules/custom_ka/barrels.dm:
```diff
@@ -0,0 +1,96 @@
+/obj/item/custom_ka_upgrade/barrels/illegal
+	name = "laser KA power converter"
+	build_name = "'Syndicate''"
+	desc = "A laser cystal ripped from a laser rifle and repurposed for kinetic accelerator assemblies."
```
Crystal.

LordFowl commented on code/modules/custom_ka/core.dm:
```diff
@@ -0,0 +1,497 @@
+
+	if(installed_barrel)
+		if(custom_name)
+			to_chat(user,"[custom_name] is written crudely in pen across the side, with the offical designation \"[official_name]\" is etched neatly on the side.")
```
Crude grammar. `While the official designation X is etched neatly...` or `With the official designation X etched neatly...`

LordFowl commented on code/modules/custom_ka/upgrade_chips.dm:
```diff
@@ -0,0 +1,64 @@
+
+/obj/item/custom_ka_upgrade/upgrade_chips/focusing
+	name = "upgrade chip - focusing"
+	desc = "Increases damage and accuracy, however reduces the range and reduce aoe capabilities."
```
Reduces.

LordFowl commented on code/modules/mining/machine_rigpress.dm:
```diff
@@ -19,8 +19,10 @@
-		var/list/kinetic_mods = list()
-		var/kineticaccelerator
+
+		//var/list/kinetic_mods = list()
```
Remove, don't comment out.

LordFowl commented on code/modules/mining/machine_rigpress.dm:
```diff
@@ -33,12 +35,13 @@
-			kineticaccelerator = 1
-			for(var/obj/item/borg/upgrade/modkit/kmod in KA.modkits)
-				kinetic_mods += kmod
+		//TODO: MOUNTED CUSTOM KAS
```
Likewise. Too many abandoned todos in this codebase to count.

LordFowl commented on code/modules/mining/machine_rigpress.dm:
```diff
@@ -61,13 +64,17 @@
-				else
-					new outcome_path(src.loc)
+
+				//TODO: MOUNTED CUSTOM KAS
```
Likewise.

LordFowl commented on html/changelogs/burgerbb-ka.yml:
```diff
@@ -0,0 +1,39 @@
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - rscadd: "Removed basic kinetic accelerators from code. Replaced them with customizable kinetic accelerators with a robust array of customization. Weaker custom kinetic accelerators can be purchased from the mining vendor or found randomly in cargo, while the stronger variants can be researched and produced by science."
+  - rscadd: "Added traitor kinetic accelerators to uplink, as well as one in the merc headquarters. These ones shoot laser beams, and can accept custom accelerator parts."
```
This is no longer true.

BurgerLUA commented:
@LordFowl The level 1 kinetic accelerator was accidentally made powerful due to an oversight in code and an attempt to make all modules "equal but different". It fires significantly slower than the original KA, however. I also hardcapped the damage per projectile to living targets to 30 to avoid powercreep.

Recoil based reloading takes away recoil from the gun because it converts that recoil into kinetic energy, effectively removing/reducing it.

As for the snowflake implementation, there are plans to merge the two systems together in a future PR

LordFowl commented:
I'm not really keen on the "I'll fix it later" conditional, but I will in this incident defer to an additional judgement.

LordFowl commented:
I personally don't think merging a "faulty" (in the sense that it is redundant, not necessarily poor) system under the promise of "fixing" (e.g. merging the two systems to not be redundant) later is a good practice.

BurgerLUA commented:
@LordFowl I am making a whole bunch of new sprites and whatnot for a future update. I will keep that feedback in mind, as well as future feedback on the current KAs, when implementing the changes,
